src/java/org/apache/cassandra/db/compaction/unified/ShardedMultiWriter.java

```diff
@@ -76,121 +75,111 @@ public ShardedMultiWriter(CompactionRealm realm,
-        this.minSstableSizeInBytes = minSstableSizeInBytes;
         this.boundaries = boundaries;
-        this.writers = new SSTableMultiWriter[boundaries.size()];
-        this.estimatedSSTables = (int) Math.max(1, Math.ceil(realm.metrics().flushSizeOnDisk().get() / minSstableSizeInBytes));
+        this.writers = new SSTableWriter[this.boundaries.count()]; // at least one
 
-        this.currentBoundary = 0;
         this.currentWriter = 0;
         this.writers[currentWriter] = createWriter(descriptor);
     }
 
-    private SSTableMultiWriter createWriter()
+    private SSTableWriter createWriter()
     {
         Descriptor newDesc = realm.newSSTableDescriptor(descriptor.directory);
         return createWriter(newDesc);
     }
 
-    private SSTableMultiWriter createWriter(Descriptor desc)
+    private SSTableWriter createWriter(Descriptor desc)
     {
-        return SimpleSSTableMultiWriter.create(desc,
-                                               forSplittingKeysBy(estimatedSSTables),
-                                               repairedAt,
-                                               pendingRepair,
-                                               isTransient,
-                                               realm.metadataRef(),
-                                               meta,
-                                               header,
-                                               indexGroups,
-                                               lifecycleNewTracker);
+        return SSTableWriter.create(desc,
+                                    forSplittingKeysBy(boundaries.count()),
+                                    repairedAt,
+                                    pendingRepair,
+                                    isTransient,
+                                    realm.metadataRef(),
+                                    new MetadataCollector(realm.metadata().comparator).commitLogIntervals(commitLogPositions),
+                                    header,
+                                    indexGroups,
+                                    lifecycleNewTracker);
     }
 
     private long forSplittingKeysBy(long splits) {
         return splits <= 1 ? keyCount : keyCount / splits;
     }
 
     @Override
-    public boolean append(UnfilteredRowIterator partition)
+    public void append(UnfilteredRowIterator partition)
     {
         DecoratedKey key = partition.partitionKey();
 
-        boolean boundaryCrossed = false;
-        /*
-        The comparison to detect a boundary is costly, but if we only do this when the size is above the threshold,
-        we may detect a boundary change in the middle of a shard and split sstables at the wrong place.
-         */
-        while (currentBoundary < boundaries.size() && key.compareTo(boundaries.get(currentBoundary)) >= 0)
+        if (boundaries.advanceTo(key.getToken())
+            && (currentWriter != 0 || writers[0].getFilePointer() > 0))
```

Comment:
Because otherwise we would try to create an empty sstable for all shards that we skip over. The extra condition is now just `uncompressedSize > 0` (the previous one worked by virtue of this only being used to flush, which is always a single range).